src/models/OrGuard.java

```diff
@@ -119,4 +119,18 @@ public String toString() {
     public int hashCode() {
         return Objects.hash(false);
     }
+
+    @Override
+    public String prettyPrint() {
```

Comment:
The `prettyPrint()` of the `OrGuard` and the `AndGuard` are exactly the same except the printed sign `&&` or `||`. Can you generalize the method, maybe in the parent `Guard` class?

Reply:
While fixing this I thought that the string returned from `toString` would suffice, though with the "drawback" of putting parenthesis around the entire guard, turning
> [ y-x<=0 && x-x<=0 && y-x<=0 && x-y<=0 && x-y<=0 && y-y<=0 && x-x<=0 && y-x<=0 && y-x<=0 && x-y<=0 && y-y<=0 && x-y<=0 ]

into

> [ (y-x<=0 && x-x<=0 && y-x<=0 && x-y<=0 && x-y<=0 && y-y<=0 && x-x<=0 && y-x<=0 && y-x<=0 && x-y<=0 && y-y<=0 && x-y<=0) ]

Is this okay, or should I revert `prettyPrint` or change the implementation for `toString`?

Reply:
I think you misunderstood me. The output produced was completely fine; I was more concerned about having (almost) duplicate code. 

If you compare the code of `OrGuard.prettyPrint()` and `AndGuard.prettyPrint()`, you see that the code is almost the same except for the printed connector symbol `||` vs `&&`, respectively. From the point of maintaining the code, it looks like it would be better to have it in once place where, depending on whom is calling is, it either prints `||` or `&&`. Maybe we can place this in the parent class `Guard`, or somewhere as a static helper function.

We could have something like
```Java
public String prettyPrintCommon(string connector) {
    return guards.stream()
            .limit(guards.size()-1)
            .map(g -> {
                if (g instanceof OrGuard || g instanceof AndGuard)
                    return "(" + g.prettyPrint() + ") " + connector + " ";
                else
                    return g.prettyPrint() + " " + connecotr + " ";
            })
            .collect(Collectors.joining())
            + guards.get(guards.size()-1).prettyPrint();
}
```
where `OrGuard` would have
```Java
public String prettyPrint() {
    return prettyPrintCommon("||")
}
```
and `AndGuard` would have
```Java
public String prettyPrint() {
    return prettyPrintCommon("&&")
}
```

Reply:
Above, the string concatination looks horible. Maybe we can instead use string formatter. 

For example
```java
return "(" + g.prettyPrint() + ") " + connector + " ";
```
would become
```java
return String.format("(%s) %s ", g.prettyPrint(), connector);
```

Reply:
I agree that a `prettyPrintCommon(string connector)` would be nice, but with that implementation `Guard` would need the `List<Guard> guards` field, which would be empty on the other derivatives (`ClockGuard`, `BoolGuard`, etc.). A potential solution could be to have `static String prettyPrintCommon(string connector, List<Guard> guards)`. Which do you prefer, or do you have a different solution?

Reply:
Well spotted that we also need the `List<Guard> guards`. I'm also fine with method overloading (something that is done frequently in Java) if needed.

BTW, as the other derivatives have no children (as I understand from your comment), maybe they can just rely on the `prettyPrint` of the parent `Guard` class, and only `OrGuard` and `AndGuard` refer to this commen function called with `connector` and `guards`.

Reply:
I have made a new function for it now (`compositePrettyPrint`). The default implementation is to return `toString` meaning only `AndGuard` and `OrGuard` need to override `prettyPrint`, and their implementation is simply `return Guard.compositePrettyPrint(guards, "&&")` and `return Guard.compositePrettyPrint(guards, "||")` respectively.